### PR TITLE
Add initializer list constructor to struct/union typet and cleanup API use [blocks: #3597]

### PR DESCRIPTION
**Pull request by tautschnig**
Use the rvalue constructor and also review any other use of constructors
to avoid unnecessary copies.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: unit/pointer-analysis/value_set.cpp [hunk 1]
```diff
@@ -47,29 +47,24 @@ SCENARIO(
   {
     // Create struct A { A *x; A *y }
 
-    struct_typet struct_A;
+    struct_typet struct_A({{"x", pointer_typet(symbol_typet("A"), 64)},
+                           {"y", pointer_typet(symbol_typet("A"), 64)}});
     struct_A.set_tag("A");
```
Comment (kroening) [inline unit/pointer-analysis/value_set.cpp:52]: These should now actually be struct_tag types!

File: src/cpp/cpp_typecheck_expr.cpp [hunk 2]
```diff
@@ -1182,7 +1182,7 @@ void cpp_typecheckt::typecheck_expr_member(
       else
       {
         // it must be a static component
-        const struct_typet::componentt pcomp=
+        const struct_typet::componentt &pcomp =
           type.get_component(to_symbol_expr(symbol_expr).get_identifier());
```
Comment (kroening) [inline src/cpp/cpp_typecheck_expr.cpp:1179]: Seems unrelated to PR?

File: src/cpp/cpp_typecheck_resolve.cpp [hunk 3]
```diff
@@ -2210,10 +2210,9 @@ void cpp_typecheck_resolvet::filter_for_named_scopes(
...
       if(id.is_member)
       {
-        struct_typet struct_type=
-          static_cast<const struct_typet &>(
-            cpp_typecheck.lookup(id.class_identifier).type);
-        const exprt pcomp=struct_type.get_component(identifier);
+        const struct_typet &struct_type = static_cast<const struct_typet &>(
+          cpp_typecheck.lookup(id.class_identifier).type);
+        const exprt &pcomp = struct_type.get_component(identifier);
         assert(pcomp.is_not_nil());
```
Comment (kroening) [inline src/cpp/cpp_typecheck_resolve.cpp]: Same here?

File: src/ansi-c/c_typecheck_expr.cpp [hunk 4]
```diff
@@ -1557,7 +1557,7 @@ void c_typecheck_baset::typecheck_expr_member(exprt &expr)
     expr.get(ID_component_name);
 
   // first try to find directly
-  struct_union_typet::componentt component=
+  const struct_union_typet::componentt &component =
     struct_union_type.get_component(component_name);
```
Comment (kroening) [inline src/ansi-c/c_typecheck_expr.cpp:1543]: Seems unrelated?

File: unit/pointer-analysis/value_set.cpp (hunk 1 above)
Comment (tautschnig) [reply unit/pointer-analysis/value_set.cpp:52]: If that's a strict requirement in goto programs then we need a validation pass that says so?! Otherwise that unit test would seem ok to me?

File: src/cpp/cpp_typecheck_expr.cpp (hunk 2 above)
Comment (tautschnig) [reply src/cpp/cpp_typecheck_expr.cpp:1179]: I may have overstretched the "cleanup API use."

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 2133fbc24da1075af6dd1d01270061d4113b8ad8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95909725

File: src/cpp/cpp_typecheck_conversions.cpp [hunk 5]
```diff
@@ -609,13 +609,11 @@ bool cpp_typecheckt::standard_conversion_pointer_to_member(
     return true;
   }
 
-  struct_typet from_struct =
-    to_struct_type(follow(static_cast<const typet &>
-      (expr.type().find("to-member"))));
+  const struct_typet &from_struct = to_struct_type(
+    follow(static_cast<const typet &>(expr.type().find("to-member"))));
```
Comment (peterschrammel) [inline src/cpp/cpp_typecheck_conversions.cpp]: ⛏ Define an irep_id for it?

**Review by peterschrammel — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4bda3d9e42b002a1045003d2c1bdaf2791f79b45).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/96062860

File: src/cpp/cpp_typecheck_conversions.cpp (hunk 5 above)
Comment (tautschnig) [reply src/cpp/cpp_typecheck_conversions.cpp]: Done in #3658.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 71045cd69516257ef6de58017c156a1f3e284042).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/96377847

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 66e000c8b4a4a99c052cec3ea1311137068a648a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/96875958

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 22d614d9b6a3b2cc7c84fcd563cf7a094fdb79a6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97216287

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: ac4ae98a8c869dc09120641a9e3625f690a2e336).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97321137

**Review by kroening — APPROVED** [review]
